From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MasterDash5 <github@masterdash5.com>
Date: Wed, 20 Aug 2025 16:29:45 -0600
Subject: Adventure API


diff --git a/src/main/java/com/hpfxd/pandaspigot/adventure/VanillaChatMessageLogic.java b/src/main/java/com/hpfxd/pandaspigot/adventure/VanillaChatMessageLogic.java
index 658d8e4d3..5fc9e8ada 100644
--- a/src/main/java/com/hpfxd/pandaspigot/adventure/VanillaChatMessageLogic.java
+++ b/src/main/java/com/hpfxd/pandaspigot/adventure/VanillaChatMessageLogic.java
@@ -1,4 +1,46 @@
 package com.hpfxd.pandaspigot.adventure;

+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.ComponentLike;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
+
+import java.util.function.BiFunction;
+import java.util.regex.MatchResult;
+
 public class VanillaChatMessageLogic {
+
+    public static Component formatChat(final Component displayName, final String format, final String message) {
+        class Replacement implements BiFunction<MatchResult, TextComponent.Builder, ComponentLike> {
+            private int index = 0;
+
+            @Override
+            public ComponentLike apply(MatchResult result, TextComponent.Builder builder) {
+                if (index == 0) {
+                    index++;
+                    return displayName;
+                }
+
+                if (index == 1) {
+                    index++;
+                    return LegacyComponentSerializer.legacySection().deserialize(message).mergeStyle(builder.asComponent());
+                }
+
+                return builder;
+            }
+        }
+
+        Replacement replacement = new Replacement();
+
+        if (format.contains("%2$s") && !format.contains("%1$s")) {
+            replacement.index = 1;
+        }
+
+        return LegacyComponentSerializer.legacySection().deserialize(format).replaceText(config -> {
+            config.times(2);
+            config.match("%(\\d+\\$)?s");
+            config.replacement(replacement);
+        });
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index 916aa2f6e..997f2b60a 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -279,7 +279,7 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
                             if (PlayerPreLoginEvent.getHandlerList().getRegisteredListeners().length != 0) {
                                 final PlayerPreLoginEvent event = new PlayerPreLoginEvent(playerName, address, uniqueId);
                                 if (asyncEvent.getResult() != PlayerPreLoginEvent.Result.ALLOWED) {
-                                    event.disallow(asyncEvent.getResult(), asyncEvent.getKickMessage());
+                                    event.disallow(asyncEvent.getResult(), asyncEvent.kickMessage()); // PandaSpigot - Adventure
                                 }
                                 Waitable<PlayerPreLoginEvent.Result> waitable = new Waitable<PlayerPreLoginEvent.Result>() {
                                     @Override
@@ -290,12 +290,12 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo

                                 LoginListener.this.server.processQueue.add(waitable);
                                 if (waitable.get() != PlayerPreLoginEvent.Result.ALLOWED) {
-                                    disconnect(event.getKickMessage());
+                                    disconnect(com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(event.kickMessage())); // PandaSpigot - Adventure
                                     return;
                                 }
                             } else {
                                 if (asyncEvent.getLoginResult() != AsyncPlayerPreLoginEvent.Result.ALLOWED) {
-                                    disconnect(asyncEvent.getKickMessage());
+                                    disconnect(com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(asyncEvent.kickMessage())); // PandaSpigot - Adventure
                                     return;
                                 }
                             }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index d25ec148d..8b7fe833e 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1942,7 +1942,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             this.server.getPluginManager().callEvent(event);

             if (!event.isCancelled()) {
-                System.arraycopy(org.bukkit.craftbukkit.block.CraftSign.sanitizeLines(event.getLines()), 0, tileentitysign.lines, 0, 4);
+                System.arraycopy(com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(event.lines()).toArray(), 0, tileentitysign.lines, 0, 4); // PandaSpigot - Adventure
                 tileentitysign.isEditable = false;
              }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0c6cdfed0..538f1d3f0 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1241,6 +1241,13 @@ public final class CraftServer implements Server {
         return configuration.getString("settings.shutdown-message");
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component shutdownMessage() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(getShutdownMessage());
+    }
+    // PandaSpigot end - Adventure
+
     @Override
     public int getSpawnRadius() {
         return ((DedicatedServer) console).propertyManager.getInt("spawn-protection", 16);
@@ -1607,6 +1614,19 @@ public final class CraftServer implements Server {
         return new CraftInventoryCustom(owner, size, title);
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public Inventory createInventory(InventoryHolder owner, InventoryType type, net.kyori.adventure.text.Component title) {
+        return new CraftInventoryCustom(owner, type, net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(title));
+    }
+
+    @Override
+    public Inventory createInventory(InventoryHolder owner, int size, net.kyori.adventure.text.Component title) throws IllegalArgumentException {
+        Validate.isTrue(size % 9 == 0, "Chests must have a size that is a multiple of 9!");
+        return new CraftInventoryCustom(owner, size, net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(title));
+    }
+    // PandaSpigot end - Adventure
+
     @Override
     public HelpMap getHelpMap() {
         return helpMap;
@@ -1647,6 +1667,13 @@ public final class CraftServer implements Server {
         return console.getMotd();
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component motd() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(console.getMotd());
+    }
+    // PandaSpigot end - Adventure
+
     @Override
     public WarningState getWarningState() {
         return warningState;
@@ -1832,6 +1859,13 @@ public final class CraftServer implements Server {
     }
     // PandaSpigot end

+    // PandaSpigot start - Adventure
+    @Override
+    public Iterable<? extends net.kyori.adventure.audience.Audience> audiences() {
+        return com.google.common.collect.Iterables.concat(Collections.singleton(getConsoleSender()), getOnlinePlayers());
+    }
+    // PandaSpigot end - Adventure
+
     private final Spigot spigot = new Spigot()
     {

diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
index 43adfccd7..5fcbbd9c2 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -11,48 +11,57 @@ import org.bukkit.craftbukkit.util.CraftChatMessage;

 public class CraftSign extends CraftBlockState implements Sign {
     private final TileEntitySign sign;
-    private final String[] lines;
+    private java.util.List<net.kyori.adventure.text.Component> lines; // PandaSpigot - Adventure

     public CraftSign(final Block block) {
         super(block);

         CraftWorld world = (CraftWorld) block.getWorld();
         sign = (TileEntitySign) world.getTileEntityAt(getX(), getY(), getZ());
-        // Spigot start
-        if (sign == null) {
-            lines = new String[]{"", "", "", ""};
-            return;
-        }
-        // Spigot end
-        lines = new String[sign.lines.length];
-        System.arraycopy(revertComponents(sign.lines), 0, lines, 0, lines.length);
+        lines = com.hpfxd.pandaspigot.adventure.PandaAdventure.asAdventure(com.google.common.collect.Lists.newArrayList(sign.lines)); // PandaSpigot - Adventure
     }

     public CraftSign(final Material material, final TileEntitySign te) {
         super(material);
         sign = te;
-        lines = new String[sign.lines.length];
-        System.arraycopy(revertComponents(sign.lines), 0, lines, 0, lines.length);
+        lines = com.hpfxd.pandaspigot.adventure.PandaAdventure.asAdventure(com.google.common.collect.Lists.newArrayList(sign.lines)); // PandaSpigot - Adventure
     }

     public String[] getLines() {
-        return lines;
+        return lines.stream().map(net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection()::serialize).toArray(String[]::new); // PandaSpigot - Adventure
     }

     public String getLine(int index) throws IndexOutOfBoundsException {
-        return lines[index];
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(lines.get(index));
     }

     public void setLine(int index, String line) throws IndexOutOfBoundsException {
-        lines[index] = line;
+        lines.set(index, net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(line)); // PandaSpigot - Adventure
+    }
+
+    // PandaSpigot start - Adventure
+    @Override
+    public java.util.List<net.kyori.adventure.text.Component> lines() {
+        return lines;
+    }
+
+    @Override
+    public net.kyori.adventure.text.Component line(int index) {
+        return lines.get(index);
+    }
+
+    @Override
+    public void line(int index, net.kyori.adventure.text.Component line) {
+        lines.set(index, line);
     }
+    // PandaSpigot end - Adventure

     @Override
     public boolean update(boolean force, boolean applyPhysics) {
         boolean result = super.update(force, applyPhysics);

         if (result) {
-            IChatBaseComponent[] newLines = sanitizeLines(lines);
+            IChatBaseComponent[] newLines = com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(lines).toArray(new IChatBaseComponent[0]); // PandaSpigot - Adventure
             System.arraycopy(newLines, 0, sign.lines, 0, 4);
             sign.update();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index aacee3ccd..a7c43ab19 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -66,6 +66,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.scoreboard.Scoreboard;
 // PaperSpigot start
 import org.github.paperspigot.Title;
+import org.jetbrains.annotations.NotNull;
 // PaperSpigot end

 @DelegateDeserialization(CraftOfflinePlayer.class)
@@ -181,6 +182,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().playerConnection.sendPacket(packet);
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public void sendMessage(final net.kyori.adventure.identity.Identity identity, final net.kyori.adventure.text.Component component, final net.kyori.adventure.audience.MessageType message) {
+        if (getHandle().playerConnection == null) return;
+
+        PacketPlayOutChat packet = new PacketPlayOutChat();
+        packet.adventure$message = component;
+        getHandle().playerConnection.sendPacket(packet);
+    }
+    // PandaSpigot end - Adventure
+
     @Override
     public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
         PacketPlayOutPlayerListHeaderFooter packet = new PacketPlayOutPlayerListHeaderFooter();
@@ -1707,14 +1719,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             getHandle().playerConnection.sendPacket(packet);
         }

-        // PandaSpigot start - Adventure
-        @Override
-        public void sendMessage(final net.kyori.adventure.identity.Identity identity, final net.kyori.adventure.text.Component message, final net.kyori.adventure.audience.MessageType type) {
-            PacketPlayOutChat packet = new PacketPlayOutChat(com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(message));
-            getHandle().playerConnection.sendPacket(packet);
-        }
-        // PandaSpigot end - Adventure
-
         // PaperSpigot start - Implement affects spawning API
         @Override
         public boolean getAffectsSpawning() {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index 44b3c3ac1..d7b2c19cf 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -271,6 +271,41 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         }
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component title() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(title);
+    }
+
+    @Override
+    public net.kyori.adventure.inventory.Book title(net.kyori.adventure.text.Component title) {
+        this.title = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(title);
+        return this;
+    }
+
+    @Override
+    public net.kyori.adventure.text.Component author() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(author);
+    }
+
+    @Override
+    public net.kyori.adventure.inventory.Book author(net.kyori.adventure.text.Component author) {
+        this.author = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(author);
+        return this;
+    }
+
+    @Override
+    public List<net.kyori.adventure.text.Component> pages() {
+        return com.hpfxd.pandaspigot.adventure.PandaAdventure.asAdventure(pages);
+    }
+
+    @Override
+    public net.kyori.adventure.inventory.Book pages(List<net.kyori.adventure.text.Component> pages) {
+        this.pages = com.hpfxd.pandaspigot.adventure.PandaAdventure.asVanilla(pages);
+        return this;
+    }
+    // PandaSpigot end - Adventure
+
     private boolean isValidPage(int page) {
         return page > 0 && page <= pages.size();
     }
@@ -319,7 +354,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     }

     @Override
-    Builder<String, Object> serialize(Builder<String, Object> builder) {
+    com.google.common.collect.ImmutableMap.Builder<String, Object> serialize(com.google.common.collect.ImmutableMap.Builder<String, Object> builder) { // PandaSpigot - Adventure
         super.serialize(builder);

         if (hasTitle()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
index ae3341c4e..c19545bac 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
@@ -129,7 +129,7 @@ class CraftMetaBookSigned extends CraftMetaBook implements BookMeta {
     }

     @Override
-    Builder<String, Object> serialize(Builder<String, Object> builder) {
+    com.google.common.collect.ImmutableMap.Builder<String, Object> serialize(com.google.common.collect.ImmutableMap.Builder<String, Object> builder) { // PandaSpigot - Adventure
         super.serialize(builder);
         return builder;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index eaaa6a11a..72467176e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -647,6 +647,18 @@ class CraftMetaItem implements ItemMeta, Repairable {
         this.displayName = name;
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component displayName() {
+        return com.hpfxd.pandaspigot.adventure.PandaAdventure.GSON.deserialize(displayName);
+    }
+
+    @Override
+    public void displayName(net.kyori.adventure.text.Component displayName) {
+        this.displayName = com.hpfxd.pandaspigot.adventure.PandaAdventure.GSON.serialize(displayName);
+    }
+    // PandaSpigot end - Adventure
+
     public boolean hasDisplayName() {
         return !Strings.isNullOrEmpty(displayName);
     }
@@ -760,6 +772,18 @@ class CraftMetaItem implements ItemMeta, Repairable {
         }
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public List<net.kyori.adventure.text.Component> lore() {
+        return lore.stream().map(net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection()::deserialize).collect(java.util.stream.Collectors.toList());
+    }
+
+    @Override
+    public void lore(List<net.kyori.adventure.text.Component> lore) {
+        this.lore = lore.stream().map(net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection()::serialize).collect(java.util.stream.Collectors.toList());
+    }
+    // PandaSpigot end - Adventure
+
     public int getRepairCost() {
         return repairCost;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
index 5bd8cf9d7..b95e1082c 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
@@ -43,6 +43,14 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
         objective.setDisplayName(displayName);
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component displayName() throws IllegalStateException {
+        checkState();
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(objective.getDisplayName());
+    }
+    // PandaSpigot end - Adventure
+
     public String getCriteria() throws IllegalStateException {
         CraftScoreboard scoreboard = checkState();

diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index f5ff0f712..3f70a5e07 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -69,6 +69,23 @@ final class CraftTeam extends CraftScoreboardComponent implements Team {
         team.setSuffix(suffix);
     }

+    // PandaSpigot start - Adventure
+    @Override
+    public net.kyori.adventure.text.Component displayName() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(team.getDisplayName());
+    }
+
+    @Override
+    public net.kyori.adventure.text.Component prefix() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(team.getPrefix());
+    }
+
+    @Override
+    public net.kyori.adventure.text.Component suffix() {
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(team.getSuffix());
+    }
+    // PandaSpigot end - Adventure
+
     public boolean allowFriendlyFire() throws IllegalStateException {
         CraftScoreboard scoreboard = checkState();

